zap/src/parser/reports.rs

```diff
@@ -0,0 +1,273 @@
+	pub fn severity(&self) -> Severity {
+		match self {
+			Self::LexerInvalidToken { .. } => Severity::Error,
+
+			Self::ParserUnexpectedEOF { .. } => Severity::Error,
+			Self::ParserUnexpectedToken { .. } => Severity::Error,
+			Self::ParserExtraToken { .. } => Severity::Error,
+			Self::ParserExpectedInt { .. } => Severity::Error,
+
+			Self::SemanticOversizeUnreliable { .. } => Severity::Error,
+			Self::SemanticPotentiallyOversizeUnreliable { .. } => Severity::Warning,
+			Self::SemanticInvalidRange { .. } => Severity::Error,
+			Self::SemanticEmptyEnum { .. } => Severity::Error,
+			Self::SemanticEnumTagUsed { .. } => Severity::Error,
+			Self::SemanticInvalidOptValue { .. } => Severity::Error,
+			Self::SemanticUnknownOptName { .. } => Severity::Warning,
+			Self::SemanticUnknownTypeRef { .. } => Severity::Error,
+			Self::SemanticNumOutsideRange { .. } => Severity::Error,
+		}
+	}
+
+	fn message(&self) -> String {
+		match self {
+			Self::LexerInvalidToken { .. } => "invalid token".to_string(),
+
+			Self::ParserUnexpectedEOF { expected, .. } => {
+				format!("expected {}, found end of file", expected.join(", "))
+			}
+
+			Self::ParserUnexpectedToken { expected, .. } => format!("expected {}", expected.join(", ")),
+			Self::ParserExtraToken { .. } => "extra token".to_string(),
+			Self::ParserExpectedInt { .. } => "expected integer".to_string(),
+
+			Self::SemanticOversizeUnreliable { .. } => "oversize unreliable".to_string(),
+			Self::SemanticPotentiallyOversizeUnreliable { .. } => "potentially oversize unreliable".to_string(),
+			Self::SemanticInvalidRange { .. } => "invalid range".to_string(),
+			Self::SemanticEmptyEnum { .. } => "empty enum".to_string(),
+			Self::SemanticEnumTagUsed { .. } => "enum tag used in variant".to_string(),
+			Self::SemanticInvalidOptValue { expected, .. } => format!("invalid opt value, expected {}", expected),
+			Self::SemanticUnknownOptName { .. } => "unknown opt name".to_string(),
+			Self::SemanticUnknownTypeRef { name, .. } => format!("unknown type reference '{}'", name),
+			Self::SemanticNumOutsideRange { .. } => "number outside range".to_string(),
+		}
+	}
+
+	fn code(&self) -> &str {
+		match self {
+			Self::LexerInvalidToken { .. } => "lex:001",
+
+			Self::ParserUnexpectedEOF { .. } => "parse:002",
+			Self::ParserUnexpectedToken { .. } => "parse:003",
+			Self::ParserExtraToken { .. } => "parse:004",
+			Self::ParserExpectedInt { .. } => "parse:005",
+
+			Self::SemanticOversizeUnreliable { .. } => "analyze:006",
+			Self::SemanticPotentiallyOversizeUnreliable { .. } => "analyze:007",
+			Self::SemanticInvalidRange { .. } => "analyze:008",
+			Self::SemanticEmptyEnum { .. } => "analyze:009",
+			Self::SemanticEnumTagUsed { .. } => "analyze:010",
+			Self::SemanticInvalidOptValue { .. } => "analyze:011",
+			Self::SemanticUnknownOptName { .. } => "analyze:012",
+			Self::SemanticUnknownTypeRef { .. } => "analyze:013",
+			Self::SemanticNumOutsideRange { .. } => "analyze:014",
+		}
```

Comment:
Should we give these just num codes instead to be eaisier to search on the docs?

Reply:
I think the code prefix gives some clarity to the user as to what the code pertains to - and also gives us the ability to add more codes to specific sections only. If we don't use `lex`, `parse`, and `analyze`, then I'd like to prefix with `1xxx`, `2xxx`, and `3xxx` instead.